core/src/main/java/org/elasticsearch/index/cache/IndexCacheModule.java

```diff
@@ -35,24 +35,26 @@
     // for test purposes only
     public static final String QUERY_CACHE_EVERYTHING = "index.queries.cache.everything";
 
-    private final Settings settings;
+    private Class<? extends QueryCache> queryCacheClass;
 
     public IndexCacheModule(Settings settings) {
-        this.settings = settings;
-    }
-
-    @Override
-    protected void configure() {
         String queryCacheType = settings.get(QUERY_CACHE_TYPE, INDEX_QUERY_CACHE);
-        Class<? extends QueryCache> queryCacheImpl;
         if (queryCacheType.equals(INDEX_QUERY_CACHE)) {
-            queryCacheImpl = IndexQueryCache.class;
+            queryCacheClass = IndexQueryCache.class;
         } else if (queryCacheType.equals(NONE_QUERY_CACHE)) {
-            queryCacheImpl = NoneQueryCache.class;
+            queryCacheClass = NoneQueryCache.class;
         } else {
             throw new IllegalArgumentException("Unknown QueryCache type [" + queryCacheType + "]");
         }
-        bind(QueryCache.class).to(queryCacheImpl).in(Scopes.SINGLETON);
+    }
+
+    public void setQueryCacheClass(Class<? extends QueryCache> queryCacheClass) {
```

Comment:
Except the current implementation I see for these type of things does a complete override, not wrapping (and how would you wrap a class for binding? you need an instance?). If we want to allow custom filtering on top of setting a custom cache, then we should have the necessary extension points to customize or configure the cache for this behavior.